### ISSUE TITLE
symmetric versions of (some) hints

### ISSUE THREAD

**Issue by jvoigtlaender** [ndmitchell/hlint#652]
Surrounding `Eq` and `Ord` operators, there are some hints that exist in symmetric forms, e.g.:
```yaml
    - warn: {lhs: x == True, rhs: x, name: Redundant ==}
    - hint: {lhs: x == False, rhs: not x, name: Redundant ==}
    - warn: {lhs: True == a, rhs: a, name: Redundant ==}
    - hint: {lhs: False == a, rhs: not a, name: Redundant ==}
    - warn: {lhs: x == Nothing , rhs:  isNothing x}
    - warn: {lhs: Nothing == x , rhs:  isNothing x}
    - warn: {lhs: not (a >  b), rhs: a <= b, note: incorrect if either value is NaN}
    - warn: {lhs: not (a >= b), rhs: a <  b, note: incorrect if either value is NaN}
    - warn: {lhs: not (a <  b), rhs: a >= b, note: incorrect if either value is NaN}
    - warn: {lhs: not (a <= b), rhs: a >  b, note: incorrect if either value is NaN}
```
And some that don't, e.g.:
```yaml
    - warn: {lhs: compare x y /= GT, rhs: x <= y}
    - warn: {lhs: compare x y == LT, rhs: x < y}
    - warn: {lhs: compare x y /= LT, rhs: x >= y}
    - warn: {lhs: compare x y == GT, rhs: x > y}
    - warn: {lhs: compare x y == EQ, rhs: x == y}
    - warn: {lhs: compare x y /= EQ, rhs: x /= y}
    - hint: {lhs: length x == 0, rhs: null x, note: IncreasesLaziness}
```
For example, there is currently no rule matching on `0 == length x`.

I do find that students write stuff like `0 == length x`, for which I would like to give the `Use null` warning. Experienced programmers may be more likely to tend towards writing expressions like `function argument == value` instead of `value == function argument` in conditionals, which might explain why `length x == 0` is present in the hint database, but not the converse.

Anyway, short of introducing a general mechanism for specifying hints like
```yaml
    - warn: {lhs: x == True, rhs: x, name: Redundant ==}
    - hint: {lhs: x == False, rhs: not x, name: Redundant ==}
    - warn: {lhs: x == Nothing , rhs:  isNothing x}
    - warn: {lhs: not (a >  b), rhs: a <= b, note: incorrect if either value is NaN}
    - warn: {lhs: not (a >= b), rhs: a < b, note: incorrect if either value is NaN}
```
only once and having the converse versions be added implicitly:
```yaml
    - warn: {lhs: True == x, rhs: x, name: Redundant ==}
    - hint: {lhs: False == x, rhs: not x, name: Redundant ==}
    - warn: {lhs: Nothing == x , rhs:  isNothing x}
    - warn: {lhs: not (b <  a), rhs: a <= b, note: incorrect if either value is NaN}
    - warn: {lhs: not (b <= a), rhs: a < b, note: incorrect if either value is NaN}
```
(which would be related to issue https://github.com/ndmitchell/hlint/issues/359), it might be beneficial immediately to add converse versions of some hints in an ad-hoc fashion.

This is what I propose to do.

I wouldn't do it for all hints currently in existence, for example I wouldn't add versions
```yaml
    - warn: {lhs: GT /= compare x y, rhs: x <= y}
```
etc. (since that's not something I've seen in student code yet).

Also, given the above assumption about beginners vs. more experiences programmers, I could first add them only to the `teaching` group introduced in https://github.com/ndmitchell/hlint/pull/650, and it could later be decided whether to make some of them default as well.

Here is, from my current experience, a list of symmetric versions of existing hints that I would add:
```yaml
    - hint: {lhs: "[] /= x", rhs: not (null x), name: Use null}
    - hint: {lhs: "[] == x", rhs: null x}
    - hint: {lhs: 0 <= length x, rhs: "True", name: Length always non-negative}
    - hint: {lhs: 0 == length x, rhs: null x, note: IncreasesLaziness}
    - hint: {lhs: 0 >= length x, rhs: null x, note: IncreasesLaziness}
    - hint: {lhs: 1 > length x, rhs: null x, note: IncreasesLaziness}
    - hint: {lhs: 0 /= length x, rhs: not (null x), note: IncreasesLaziness, name: Use null}
    - hint: {lhs: 0 < length x, rhs: not (null x), note: IncreasesLaziness, name: Use null}
    - hint: {lhs: 1 <= length x, rhs: not (null x), note: IncreasesLaziness, name: Use null}
```

Should I make a pull request along these lines?

**Comment by ndmitchell** [comment]
As long as its not a huge number, I'm happy to add the symmetric ones (a huge number may have an impact on runtime speed and certainly has an impact on hint maintenance). Having 10-20 seems fine, so yes please, PR away!
